clippy_lints/src/misc.rs

```diff
@@ -257,46 +256,65 @@ impl<'tcx> LateLintPass<'tcx> for LintPass {
-            // Don't lint things expanded by #[derive(...)], etc or `await` desugaring
+        if in_external_macro(cx.sess(), expr.span)
+            || expr.span.desugaring_kind().is_some()
+            || any_parent_is_automatically_derived(cx.tcx, expr.hir_id)
+        {
             return;
         }
-        let sym;
-        let binding = match expr.kind {
-            ExprKind::Path(ref qpath) if !matches!(qpath, hir::QPath::LangItem(..)) => {
-                let binding = last_path_segment(qpath).ident.as_str();
-                if binding.starts_with('_') &&
-                    !binding.starts_with("__") &&
-                    binding != "_result" && // FIXME: #944
-                    is_used(cx, expr) &&
-                    // don't lint if the declaration is in a macro
-                    non_macro_local(cx, cx.qpath_res(qpath, expr.hir_id))
+        let (definition_hir_id, ident) = match expr.kind {
+            ExprKind::Path(ref qpath) => {
+                if let QPath::Resolved(None, path) = qpath
+                    && let Res::Local(id) = path.res
+                    && is_used(cx, expr)
                 {
-                    Some(binding)
+                    (id, last_path_segment(qpath).ident)
                 } else {
-                    None
+                    return;
                 }
             },
-            ExprKind::Field(_, ident) => {
-                sym = ident.name;
-                let name = sym.as_str();
-                if name.starts_with('_') && !name.starts_with("__") {
-                    Some(name)
+            ExprKind::Field(recv, ident) => {
+                if let Some(adt_def) = cx.typeck_results().expr_ty_adjusted(recv).ty_adt_def()
+                    && let Some(field) = adt_def.all_fields().find(|field| field.name == ident.name)
+                    && let Some(local_did) = field.did.as_local()
+                    && let Some(hir_id) = cx.tcx.opt_local_def_id_to_hir_id(local_did)
+                    && !cx.tcx.type_of(field.did).skip_binder().is_phantom_data()
+                {
+                    (hir_id, ident)
                 } else {
-                    None
+                    return;
                 }
             },
-            _ => None,
+            _ => return,
         };
-        if let Some(binding) = binding {
-            span_lint(
+
+        let name = ident.name.as_str();
+        if name.starts_with('_')
+            && !name.starts_with("__")
+            && let definition_span = cx.tcx.hir().span(definition_hir_id)
+            && !definition_span.from_expansion()
+        {
+            let (level, _) = cx.tcx.lint_level_at_node(USED_UNDERSCORE_BINDING, definition_hir_id);
+            match level {
+                Level::Allow => return,
+                Level::Expect(id) => {
+                    cx.fulfill_expectation(id);
+                    return;
```

Comment:
This check might miss a second expectation, if there is also one on the statement level. I think in 99.9% of the cases it should be fine, but it might be safer to also check the level at `expr` and potentially fulfill a second expectation. :)

Otherwise very nice check :)

Reply:
Aha makes sense, does the `ForceWarn` expectation ever need to be fulfilled?

Reply:
The force warn expectation should be handled exactly the same as a normal one. That one becomes relevant if there is an `#[expect]` and the user specifies a `--force-warn` in via the console arguments

Reply:
Am I understanding correctly that `--force-warn foo` turns all the `#[expect(foo)]`s into `Level::ForceWarn(Some(_))`?

Personally I would still expect the lint to emit if `--force-warn` was passed, but it could probably still fulfill the expectation to not create a new warning/error

Reply:
Yeah, sorry, I typed my answer on the go :sweat_smile:. With `--force-warn` the lint should be emitted as normal, that will emit the lint and fulfill the lint expectation IIRC